query-translator/src/types.rs

```diff
@@ -0,0 +1,271 @@
+struct Op(String);      // TODO
+enum Constraint {
+    Infix {
+        op: Op,
+        left: ColumnOrExpression,
+        right: ColumnOrExpression
+    }
+}
+
+enum JoinOp {
+    Inner,
+}
+
+// Short-hand for a list of tables all inner-joined.
+struct TableList(Vec<SourceAlias>);
+
+struct Join {
+    left: TableOrSubquery,
+    op: JoinOp,
+    right: TableOrSubquery,
+    // TODO: constraints (ON, USING).
+}
+enum TableOrSubquery {
+    Table(SourceAlias),
+    // TODO: Subquery.
+}
+
+enum FromClause {
+    TableList(TableList),      // Short-hand for a pile of inner joins.
+    Join(Join),
+}
+
+struct SelectQuery {
+    projection: Vec<Projection>,
+    from: FromClause,
+    constraints: Vec<Constraint>,
+}
+
+//---------------------------------------------------------
+// Turn that representation into SQL.
+
+impl QueryFragment for ColumnOrExpression {
+    fn to_sql(&self, out: &mut QueryBuilder) -> BuildQueryResult {
+        use self::ColumnOrExpression::*;
+        match self {
+            &Column(QualifiedAlias(ref table, ref column)) => {
+                out.push_identifier(table.as_str())?;
+                out.push_sql(".");
+                out.push_identifier(column.as_str())
+            },
+            &Integer(i) => {
+                out.push_sql(i.to_string().as_str());
+                Ok(())
+            }
+        }
+    }
+}
+
+impl QueryFragment for Projection {
+    fn to_sql(&self, out: &mut QueryBuilder) -> BuildQueryResult {
+        self.0.to_sql(out)?;
+        out.push_sql(" AS ");
+        out.push_identifier(self.1.as_str())
+    }
+}
+
+impl QueryFragment for Op {
+    fn to_sql(&self, out: &mut QueryBuilder) -> BuildQueryResult {
+        // No escaping needed.
+        out.push_sql(self.0.as_str());
+        Ok(())
+    }
+}
+
+impl QueryFragment for Constraint {
+    fn to_sql(&self, out: &mut QueryBuilder) -> BuildQueryResult {
+        use self::Constraint::*;
+        match self {
+            &Infix { ref op, ref left, ref right } => {
+                left.to_sql(out)?;
+                out.push_sql(" ");
+                op.to_sql(out)?;
+                out.push_sql(" ");
+                right.to_sql(out)
+            }
+        }
+    }
+}
+
+impl QueryFragment for JoinOp {
+    fn to_sql(&self, out: &mut QueryBuilder) -> BuildQueryResult {
+        out.push_sql(" JOIN ");
+        Ok(())
+    }
+}
+
+// We don't own SourceAlias or QueryFragment, so we can't implement the trait.
+fn source_alias_to_sql(out: &mut QueryBuilder, sa: &SourceAlias) -> BuildQueryResult {
+    let &SourceAlias(ref table, ref alias) = sa;
+    out.push_identifier(table.name())?;
+    out.push_sql(" AS ");
+    out.push_identifier(alias.as_str())
+}
+
+impl QueryFragment for TableList {
+    fn to_sql(&self, out: &mut QueryBuilder) -> BuildQueryResult {
+        if self.0.is_empty() {
+            return Ok(());
+        }
+
+        source_alias_to_sql(out, &self.0[0])?;
+
+        for sa in self.0[1..].iter() {
```

Comment:
Consider `self.0.iter.skip(1)`.  Or grab the iter and call `.next()` rather than indexing.